memcache/selector.go

```diff
@@ -32,15 +36,53 @@ type ServerSelector interface {
 	// PickServer returns the server address that a given item
 	// should be shared onto.
 	PickServer(key string) (net.Addr, error)
+	// Each iterates func over servers, returning the first non-nil error
 	Each(func(net.Addr) error) error
+	// OnResult informs the selector of the result of an operation - err is nil on success
+	OnResult(addr net.Addr, err error)
 }
 
-// ServerList is a simple ServerSelector. Its zero value is usable.
+const maxRetryWait = time.Second * 10
+const startingWait = time.Millisecond * 10
+
+// retryState drives the per connection state-machine through the retry process
```

Comment:
Can you leave a comment explaining the state changes that `retryState` goes through. It took me a bit of time to rationalize them. Something akin to:
1) `retryWait` is the starting retry state for a failed connection to a given address. The address is removed from the list of available addresses that `PickServer` can choose from.
2) A goroutine is created that waits before changing the `retryWait` state to `retryReady`. Once this happens the address is put back in `available`. 
3) If the address is picked by `PickServer` and it is in the `retryReady` state, the state moves to `retryRunning` (effectively testing the connection implicitly since we know an `OnResult` call is always made post `dial` or reading / writing from a connection). 
4) In the `retryRunning` state, future errors will bring the address back into a `retryWait` state but will also double the wait time for the scheduled goroutine mentioned in point 3.

Reply:
Absolutely, I've taken the above as it was excellent.